### PR TITLE
Enable GKE deployer to use custom subnet for the cluster

### PR DESCRIPTION
**Pull request by shyamjvs**
/cc @wojtek-t @porridge @kubernetes/sig-scalability-misc

### CODE REVIEW

**Comment by krzyzacy** [comment]
/cc @MrHohn 
I think you had a similar change, maybe that's for gce only?

or do you also need to change `--mode=auto` or `--subnet-mode` to custom when creating the network?

**Comment by MrHohn** [comment]
> I think you had a similar change, maybe that's for gce only?

Exactly, my change only took effects on gce. Great to have this on gke! Thanks @shyamjvs

File: kubetest/gke.go [hunk 1]
```diff
@@ -259,6 +261,17 @@ func (g *gkeDeployer) Up() error {
 			return err
 		}
 	}
+	// Create a custom subnet in that network if it was asked for.
```
Comment (MrHohn) [inline kubetest/gke.go]: Seems like this custom subnet will never get cleaned up? If this custom subnet already exist, should we just use it?

Comment (krzyzacy) [reply kubetest/gke.go]: I think janitor will clean up custom subnet eventually - but check resource leaking might fail

**Comment by MrHohn** [comment]
> or do you also need to change --mode=auto or --subnet-mode to custom when creating the network?

Ah, I see the difference now. This is not creating a manual/custom network, but creating a custom subnet under the existing network. No need to change network mode in this case I think.

**Comment by krzyzacy** [comment]
@MrHohn isn't `--subnet-mode` default to `AUTO`? Or it's intended to create both auto/custom subnet? (:joy: I'm a network noob)

**Comment by shyamjvs** [comment]
That's right, we're just adding a new custom subnet to an already existing
network (in auto-mode). We don't need to worry about cleaning it up iiuc.

**Comment by MrHohn** [comment]
> isn't --subnet-mode default to AUTO? Or it's intended to create both auto/custom subnet? (joy I'm a network noob)

Sorry, my wording is very confusing. My understand is an auto network will automatically create subnet under every regions while a custom network wouldn't. But user are free to add any more subnets under the auto && custom network.

**Comment by MrHohn** [comment]
Do we have logic else where to cleanup the new created subnet? If I recall correctly, deleting a network will automatically remove the auto created subnets, but not the new added ones.

For instance I added a "test" subnet to an auto network, and the network deletion failed.
```
$ gcloud compute networks subnets list --network e2e
NAME  REGION                   NETWORK  RANGE
e2e   asia-northeast1          e2e      10.146.0.0/20
e2e   us-west1                 e2e      10.138.0.0/20
e2e   southamerica-east1       e2e      10.158.0.0/20
e2e   europe-west4             e2e      10.164.0.0/20
e2e   asia-east1               e2e      10.140.0.0/20
e2e   asia-southeast1          e2e      10.148.0.0/20
e2e   us-east4                 e2e      10.150.0.0/20
e2e   europe-west1             e2e      10.132.0.0/20
e2e   europe-west2             e2e      10.154.0.0/20
e2e   europe-west3             e2e      10.156.0.0/20
e2e   australia-southeast1     e2e      10.152.0.0/20
e2e   asia-south1              e2e      10.160.0.0/20
e2e   us-east1                 e2e      10.142.0.0/20
e2e   us-central1              e2e      10.128.0.0/20
test  us-central1              e2e      10.124.0.0/20
e2e   northamerica-northeast1  e2e      10.162.0.0/20
$ gcloud compute networks delete e2e
The following networks will be deleted:
 - [e2e]

Do you want to continue (Y/n)?  y

ERROR: (gcloud.compute.networks.delete) Could not fetch resource:
 - The network resource 'projects/zihongz-k8s-dev/global/networks/e2e' is already being used by 'projects/zihongz-k8s-dev/regions/us-central1/subnetworks/test'
```

**Comment by shyamjvs** [comment]
@MrHohn Nice catch - thanks for pointing it. I've now added some cleanup logic, PTAL.

File: kubetest/gke.go [hunk 2]
```diff
@@ -514,6 +532,11 @@ func (g *gkeDeployer) Down() error {
 			g.location)...))
 	errFirewall := finishRunning(exec.Command("gcloud", "compute", "firewall-rules", "delete", "-q", firewall,
 		"--project="+g.project))
+	var errSubnet error
+	if g.subnetwork != "" {
+		errSubnet := finishRunning(exec.Command("gcloud", "compute", "networks", "subnets", "delete", "-q", g.subnetwork,
+			g.subnetworkRegion, "--project="+g.project))
```
Comment (krzyzacy) [inline kubetest/gke.go]: --region=g.subnetworkRegion?

Comment (shyamjvs) [reply kubetest/gke.go]: Nope, it's already the whole thing. See the description of the flag from which it's computed.

Comment (krzyzacy) [reply kubetest/gke.go]: got it 👍

**Comment by krzyzacy** [comment]
lgtm, will let @MrHohn apply the stamp

File: kubetest/gke.go [hunk 3]
```diff
@@ -259,6 +262,18 @@ func (g *gkeDeployer) Up() error {
 			return err
 		}
 	}
+	// Create a custom subnet in that network if it was asked for.
+	if *gkeCustomSubnet != "" {
+		customSubnetFields := strings.Fields(*gkeCustomSubnet)
+		createSubnetCommand := []string{"compute", "networks", "subnets", "create"}
+		createSubnetCommand = append(createSubnetCommand, "--project="+g.network, "--network="+g.network)
```
Comment (MrHohn) [inline kubetest/gke.go]: Humm..Should this be `"--project="+g.project`?

Comment (krzyzacy) [reply kubetest/gke.go]: :joy: you have better eyes

Comment (shyamjvs) [reply kubetest/gke.go]: Fixed.. thanks.

**Pull request merged**